src/gluonts/dataset/loader.py

```diff
@@ -37,6 +50,61 @@ def __call__(self, data, is_train):
         yield from batcher(data, self.batch_size)
 
 
+class Stack(Transformation, BaseModel):
+    def __call__(self, data, is_train):
+        for batch in data:
+            yield rows_to_columns(batch, np.array)
+
+
+def as_stacked_batches(
+    dataset: Dataset,
+    *,
+    batch_size: int,
+    output_type: Optional[Callable] = None,
+    num_batches_per_epoch: Optional[int] = None,
+    shuffle_buffer_length: Optional[int] = None,
+    field_names: Optional[list] = None,
+):
+    """
+    Prepare data in batches to be passed to a network.
+
+    Input data is collected into batches of size ``batch_size`` and then
+    columns are stacked on top of each other. In addition, the result is
+    wrapped in ``output_type`` if provided.
+
+    If ``num_batches_per_epoch`` is provided, only those number of batches are
+    effectively returned. This is especially useful for training when
+    providing a cyclic dataset.
+
+    To pseudo shuffle data, ``shuffle_buffer_length`` can be set to collect
+    inputs into a buffer first, from which we then randomly sample.
+
+    Setting ``field_names`` will only consider those columns in the input data
+    and discard all other values.
+    """
+
+    if shuffle_buffer_length:
+        dataset = PseudoShuffled(dataset, shuffle_buffer_length)
+
+    transform: Transformation = Identity()
+
+    if field_names is not None:
+        transform += SelectFields(field_names)
+
+    transform += Batch(batch_size=batch_size)
+    transform += Stack()
+
+    if output_type is not None:
+        transform += Valmap(output_type)
+
+    # Note: is_train needs to be provided but does not have an effect
+    transformed_dataset = transform.apply(dataset, is_train=True)
+    return IterableSlice(
+        transformed_dataset,
+        num_batches_per_epoch,
+    )
```

Comment:
If this creates complications (we may or may not need to `iter` the `transformed_dataset`) we could even remove this wrapping and rely on the trainer to limit how much data is taken at once, see the `limit_train_batches` option at https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.trainer.trainer.Trainer.html#pytorch_lightning.trainer.trainer.Trainer

This way we can directly return `transformed_dataset`

Reply:
I don't think this causes issues. One is free to not set `num_batches_per_epoch`.

Reply:
it prevents you from being able to `iter(transformed_dataset)` downstream

Reply:
nvm